features.go

```diff
@@ -323,32 +324,34 @@ parsefeatures:
 			s.features[tok.Name.Space] = nil
 
 			feature, ok := getFeature(tok.Name, features)
-
-			if ok && s.state&feature.Necessary == feature.Necessary && s.state&feature.Prohibited == 0 {
+			if ok {
 				req, data, err := feature.Parse(ctx, s.in.d, &tok)
 				if err != nil {
 					return nil, err
 				}
+				sf.req = sf.req || req
```

Comment:
This line was the core of the problem. If we see that a feature in the list was required we mark that it exists so that we know we're not done with negotiation. Previously, this happened after the checks to see if we're allowed to negotiate the feature right this moment on the next line. We don't care about that though, even if we can't negotiate the feature right this moment we might be able to negotiate it in the next step after negotiating some more optional features (like StartTLS in this case).